src/utils/ContainerUtils.ts

```diff
@@ -0,0 +1,89 @@
+    }
+
+    return {};
+};
+
+export class Utils {
+    static validate(props: ContainerProps & { filterNode: HTMLElement; targetListView: ListView; isModeler?: boolean}): string {
+        const widgetName = props.friendlyId;
+        const { targetListView } = props;
+        const errorMessage = [];
+        // validate filter values if filterby = attribute, then value should not be empty or "" or " ".
+        if (!props.filterNode) {
+            return `${widgetName}: unable to find a listview with to attach to`;
+        }
+
+        if (props.filterBy === "XPath" && !props.constraint) {
+            errorMessage.push(`Filter by 'XPath' requires an 'XPath contraint'`);
+        }
+        if (props.filterBy === "attribute" && !props.attribute) {
+            errorMessage.push(`Filter by 'Attribute' requires an 'Attribute'`);
+        }
+        if (props.filterBy === "attribute" && !props.attributeValue) {
+            errorMessage.push(`Filter by 'Attribute' requires an 'Attribute value`);
+        }
+        if (props.unCheckedFilterBy === "XPath" && !props.unCheckedConstraint) {
+            errorMessage.push(`Unchecked filter by 'XPath' requires an 'XPath contraint'`);
+        }
+        if (props.unCheckedFilterBy === "attribute" && !props.unCheckedAttribute) {
+            errorMessage.push(`Unchecked filter by 'Attribute' requires an 'Attribute'`);
+        }
+        if (props.unCheckedFilterBy === "attribute" && !props.unCheckedAttributeValue) {
+            errorMessage.push(`Unchecked filter by 'Attribute' requires an 'Attribute value`);
+        }
+        if (errorMessage.length) {
+            return `${widgetName} : ${errorMessage.join(", ")}`;
+        }
+
+        if (props.isModeler) {
+            return "";
+        }
+
+        const type = targetListView && targetListView.datasource.type;
+        if (type && type !== "database" && type !== "xpath") {
+            return `${widgetName}, widget is only compatible with list view data source type 'Database' and 'XPath'`;
+        }
+        if (!(targetListView && targetListView._datasource && targetListView._entity && targetListView.update)) {
+            return `${widgetName}: this Mendix version is incompatible`;
+        }
+        if (targetListView._entity && props.listviewEntity !== targetListView._entity) {
+            return `${widgetName}: supplied entity "${props.listviewEntity}" does not belong to list view data source`;
+        }
+
+        return "";
+    }
+
+    static findTargetNode(filterNode: HTMLElement): HTMLElement | null {
+        let targetNode: HTMLElement | null = null ;
+
+        while (!targetNode && filterNode) {
+            targetNode = filterNode.querySelectorAll(`.mx-listview`)[0] as HTMLElement;
+            if (targetNode || filterNode.isEqualNode(document) || filterNode.classList.contains("mx-incubator")
+            || filterNode.classList.contains("mx-offscreen")) {
```

Comment:
Wrong indenting

Reply:
Gatcha

Reply:
?? dont get it.
Check this guideline http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-136091.html#248

Line wrapping for if statements should generally use the 8-space rule, since conventional (4 space) indentation makes seeing the body difficult. For example:

``` java
//DON'T USE THIS INDENTATION
if ((condition1 && condition2)
    || (condition3 && condition4)
    ||!(condition5 && condition6)) { //BAD WRAPS
    doSomethingAboutIt();            //MAKE THIS LINE EASY TO MISS
} 

//USE THIS INDENTATION INSTEAD
if ((condition1 && condition2)
        || (condition3 && condition4)
        ||!(condition5 && condition6)) {
    doSomethingAboutIt();
} 

//OR USE THIS
if ((condition1 && condition2) || (condition3 && condition4)
        ||!(condition5 && condition6)) {
    doSomethingAboutIt();
} 
```